util/app/path/path.go

```diff
@@ -27,3 +29,62 @@ func Path(root, path string) (string, error) {
 	}
 	return appPath, nil
 }
+
+type OutOfBoundsSymlinkError struct {
+	File string
+	Err  error
+}
+
+func (e *OutOfBoundsSymlinkError) Error() string {
+	return "out of bounds symlink found"
+}
+
+// CheckOutOfBoundsSymlinks determines if basePath contains any symlinks that
+// are absolute or point to a path outside of the basePath. If found, an
+// OutOfBoundsSymlinkError is returned.
+func CheckOutOfBoundsSymlinks(basePath string) error {
+	absBasePath, err := filepath.Abs(basePath)
+	if err != nil {
+		return fmt.Errorf("failed to get absolute path: %v", err)
+	}
+	return filepath.Walk(absBasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk for symlinks in %s: %v", absBasePath, err)
+		}
+		if files.IsSymlink(info) {
+			// We don't use filepath.EvalSymlinks because it fails without returning a path
+			// if the target doesn't exist.
+			linkTarget, err := os.Readlink(path)
+			if err != nil {
+				return fmt.Errorf("failed to read link %s: %v", path, err)
+			}
+			// get the path of the symlink relative to basePath, used for error description
+			linkRelPath, err := filepath.Rel(absBasePath, path)
+			if err != nil {
+				return fmt.Errorf("failed to get relative path for symlink: %v", err)
+			}
+			// deny all absolute symlinks
+			if filepath.IsAbs(linkTarget) {
+				return &OutOfBoundsSymlinkError{File: linkRelPath}
+			}
+			// get the parent directory of the symlink
+			currentDir := filepath.Dir(path)
+
+			// walk each part of the symlink target to make sure it never leaves basePath
+			parts := strings.Split(linkTarget, string(os.PathSeparator))
```

Comment:
I think you could just use filepath.SplitList here. The only difference should be that if the link is an empty string, `parts` will be `[]` instead of `[""]`.

Reply:
Changing it, for some reason or another, broke bad link test 3.